### ISSUE TITLE
Consider using ProjFS instead of Dokan for virtual drives

### ISSUE THREAD

**Issue by BobVul** [nextcloud/desktop#1104]
This is possibly a long-term issue, as ProjFS is still very new (optional feature in Win10 v1809). Maybe better in a year's time, but I bring it up now because migrating would only become more difficult as the current Dokan approach matures.

[ProjFS](https://docs.microsoft.com/en-us/windows/desktop/projfs/projected-file-system) is more or less Microsoft's version of FUSE. It's effectively a first-party supported version of what Dokan does.

Pros:

* First-party support from Microsoft, no need to install/update a third-party driver
    * Therefore you don't need to ship kernel-mode components for Windows
* Should work by default within a folder, no need to have a separate drive (letter)
* Used by VFSForGit, a fairly large project

Cons:

* Requires a recent version of Windows (Win10 v1809, so no LTSB support yet) -- possibly not a huge issue because virtual drives are very new to NextCloud too
* No real users other than Windows/VFSForGit at the moment
    * Though the [VFSForGit](https://github.com/Microsoft/VFSForGit) project does provide a kext for macOS. Again you have the issue of it not being first-party supported.
    * There's also a [prototype](https://github.com/github/libprojfs) for ProjFS on top of FUSE for Linux.
* Installation/activation of ProjFS currently requires enabling a Windows optional feature, i.e. it's not enabled by default yet
* This would likely be a major change. I don't know how much Dokan code could be ported, or how big the effort would be.

**Comment by WindowsNT** [comment]
ProjFs =  read only.

**Comment by BobVul** [comment]
@WindowsNT How so? The RegFS sample is, yes, because it doesn't support (blocks) modifications/deletes/renames. But [the notifications](https://docs.microsoft.com/en-us/windows/desktop/projfs/file-system-operation-notifications#receiving-notifications) ([apiref](https://docs.microsoft.com/en-us/windows/desktop/api/projectedfslib/ne-projectedfslib-prj_notification) for modification exist. For example, it's possible to register to receive a `PRJ_NOTIFICATION_FILE_HANDLE_CLOSED_FILE_MODIFIED` notification, at which point you can open the *full file*, read changes, write them to the remote storage and then (if desired) reset the file to a hydrated placeholder.

**Comment by WindowsNT** [comment]
That is correct, but only after the file finishes, You do not get real time notifications of writing like dokan.

**Comment by BobVul** [comment]
Ah, ok. If that's a requirement, fair enough.

**Comment by jgallucci32** [comment]
Does anyone know what the Gladinet/CentreStack desktop client is using under the hood? We have been using it for over 4 years now and haven't had any issues. It supports virtual folder (drive letter), custom branding at runtime (pulls config and custom favicon from server) and supports right-click context with logo. It essentially does everything the Next cloud client is trying to do, so looking there might be a good start.

**Comment by maxstoyanov** [comment]
Quote from the [official Projected Filesystem docs](https://docs.microsoft.com/en-us/windows/win32/projfs/projected-file-system):

> ProjFS is designed for use with high-speed backing data stores. One of its design goals is to make the projected data appear as if it were locally present, hiding the fact that the data may be remote. As such, ProjFS doesn't provide: mechanisms for reporting progress of data recall; **indication of the online versus offline state of a file; nor other features that may be desirable when working with backing data stores that are slow.** For such scenarios, consider instead using the [Cloud Files API](https://docs.microsoft.com/en-us/windows/win32/cfapi/cloud-files-api-portal).

We have to expect that internet connections will be slow. Offline and online support also is a requirement. So, seems like the [cloud files API](https://docs.microsoft.com/en-us/windows/win32/cfapi/cloud-files-api-portal) might be more fitting.

**Comment by BobVul** [comment]
Yea, that does look more appropriate. Should that be a new issue, or should I edit this one?

**Comment by er-vin** [comment]
Doesn't apply anymore with the new code base. Please open a new issue if that shows up with latest dailies.
